tidy3d/components/grid/grid_spec.py

```diff
@@ -665,14 +787,580 @@ def _preprocessed_structures(self, structures: List[StructureType]) -> List[Stru
+    on the cross section perpendicular to layer thickness direction can be automatically identified. Subsequently,
+    mesh is snapped and refined around the corners. Mesh can also be refined and snapped around the bounds along
+    the layer thickness direction.
+
+    Note
+    ----
+
+    Corner detection is performed on a 2D plane sitting in the middle of the layer. If the layer is finite
+    along inplane axes, corners outside the bounds are discarded.
+
+    Note
+    ----
+
+    This class only takes effect when :class:`AutoGrid` is applied.
+
+    Example
+    -------
+    >>> layer_spec = LayerRefinementSpec(axis=2, center=(0,0,0), size=(td.inf, td.inf, 1))
+
+    """
+
+    axis: Axis = pd.Field(
+        ...,
+        title="Axis",
+        description="Specifies dimension of the layer normal axis (0,1,2) -> (x,y,z).",
+    )
+
+    min_steps_along_axis: Optional[pd.PositiveFloat] = pd.Field(
+        None,
+        title="Minimal Number Of Steps Along Axis",
+        description="If not ``None`` and the thickness of the layer is nonzero, set minimal "
+        "number of steps discretizing the layer thickness.",
+    )
+
+    bounds_refinement: Optional[GridRefinement] = pd.Field(
+        None,
+        title="Mesh Refinement Factor Around Layer Bounds",
+        description="If not ``None``, refine mesh around minimum and maximum positions "
+        "of the layer along axis dimension. If `min_steps_along_axis` is also specified, "
+        "refinement here is only applied if it sets a smaller grid size.",
+    )
+
+    bounds_snapping: Optional[Literal["bounds", "lower", "upper", "center"]] = pd.Field(
+        "lower",
+        title="Placing Grid Snapping Point Along Axis",
+        description="If not ``None``, enforcing grid boundaries to pass through ``lower``, "
+        "``center``, or ``upper`` position of the layer; or both ``lower`` and ``upper`` with ``bounds``.",
+    )
+
+    corner_finder: Optional[CornerFinderSpec] = pd.Field(
+        CornerFinderSpec(),
+        title="Inplane Corner Detection Specification",
+        description="Specification for inplane corner detection. Inplane mesh refinement "
+        "is based on the coordinates of those corners.",
+    )
+
+    corner_snapping: bool = pd.Field(
+        True,
+        title="Placing Grid Snapping Point At Corners",
+        description="If ``True`` and ``corner_finder`` is not ``None``, enforcing inplane "
+        "grid boundaries to pass through corners of geometries specified by ``corner_finder``.",
+    )
+
+    corner_refinement: Optional[GridRefinement] = pd.Field(
+        GridRefinement(),
+        title="Inplane Mesh Refinement Factor Around Corners",
+        description="If not ``None`` and ``corner_finder`` is not ``None``, refine mesh around "
+        "corners of geometries  specified by ``corner_finder``. "
+        "This option is similar to edge refinement if the edge is axis-aligned.",
```

Comment:
Sorry what does edge refinement refer to here?

Reply:
Grids are denser closer to the edge. Probably remove this sentence?

Reply:
Do you mean it's similar to bounds refinement? I'm just not sure what you are referring to. :)

Reply:
Imagine that you have a metallic polygon. You'll want finer meshing near the polygon edges to resolve the more rapid field variations.